### PR TITLE
adding formName to formApi

### PR DESCRIPTION
**Pull request by vospascal**
I would like to add support for name property/attribute to final-form since form name is a default property/attribute, its handy to get some context when dynamically and re-using custom field components. 
This is part one of a two part pull request the other one will be for react-final-form.

### CODE REVIEW

**Comment by vospascal** [comment]
i dont know wy travis is failing on some node versions could use some help there

File: package.json [hunk 1]
```diff
@@ -99,6 +99,7 @@
     }
   ],
   "dependencies": {
-    "@babel/runtime": "^7.3.1"
+    "@babel/runtime": "^7.3.1",
+    "settings": "^0.1.1"
```
Comment (Sharakai) [inline package.json]: What's this `settings` dependency for? I can't see it being used anywhere.

File: src/index.d.ts [hunk 2]
```diff
@@ -102,6 +105,7 @@ export interface FieldSubscription {
   valid?: boolean
   value?: boolean
   visited?: boolean
+  formName?: string
```
Comment (Sharakai) [inline src/index.d.ts]: This should be a `boolean`

File: src/index.d.ts [hunk 3]
```diff
@@ -256,6 +264,8 @@ export interface Config<FormData = object> {
   ) => object | Promise<object | undefined> | undefined | void
   validate?: (values: object) => object | Promise<object>
   validateOnBlur?: boolean
+  name?: string
+  formName?: string
```
Comment (Sharakai) [inline src/index.d.ts]: `formName` doesn't seem to be a valid config key.
You're using `name` in the form config, and aliasing it to `formName`.
Only `name` should exist here.

File: src/types.js.flow [hunk 4]
```diff
@@ -251,14 +256,17 @@ export type Config = {
...
   mutators?: { [string]: Mutator },
   onSubmit: (
     values: Object,
     form: FormApi,
     callback: ?(errors: ?Object) => ?Object
   ) => ?Object | Promise<?Object> | void,
   validate?: (values: Object) => Object | Promise<Object>,
-  validateOnBlur?: boolean
+  validateOnBlur?: boolean,
+  name?: string,
```
Comment (Sharakai) [inline src/types.js.flow]: Same here regarding the `name` and `formName` mix-up. Only `name` is a real config property.

**Comment by vospascal** [comment]
@Sharakai thnx will fix

**Comment by vospascal** [comment]
@Sharakai you got time to check out https://github.com/final-form/react-final-form/pull/443 this implements the formName

**Comment by vospascal** [comment]
@erikras do you have anytime soon to merge the requests or have a discussion about adding it?
If you need some help with the awesome library i'm willing to where possible.

**Comment by erikras** [comment]
I'm not sure I understand why Final Form needs to maintain an additional superfluous string that doesn't change. Can you explain it further?

**Comment by vospascal** [comment]
@erikras @Sharakai Thats easy this is the way to add context and pass it down while maintaining re-uasbility for instance with validation form's are the containers that hold the fields the fields and for example validation can be re-used over multiple forms. In for any component based architecture react/angular/vue this will be very helpfull. This way you dont have to resort to property drilling. Thats wy i added also an pull request to the react-final-form repo.

**Comment by erikras** [comment]
Can you give me a concrete example? I'm still not seeing the utility.

**Comment by vospascal** [comment]
The answer is context without property drilling.
Having a static named property in any case makes configuration of reusable components doable.
It could also be something else then the form name aslong as it provides context to the fields inside the form. 
I know this is wat makes it look like its superfluous, redundant, unneeded but its the only property that provides basic context. 
Any other suggestions how to achieve this are welcome.

**Comment by erikras** [comment]
_Who_ needs to know the form name? The validation function? The custom input component?

**Comment by vospascal** [comment]
@erikras that should also work indeed good idea! That i didnt think of that before.... 
Dont know if that also goes for the non react usages of final-form but this will work fine for my usecase i will implement this first thing tomorrow.

**Comment by vospascal** [comment]
Got it implemented looks like it works prefectly thnx ill close this issue.
I also got an implementation using a decorator for meta data that might be interessting. ill see if i got the time this week to wip up an example implementing an typeahead.
I hope you can take a look.

**Pull request closed without merging**